sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1063,15 +1060,14 @@ static ipv6_addr_t *_src_addr_selection(gnrc_netif_t *netif,
             return ptr;
         }
         /* Rule 2: Prefer appropriate scope. */
-        /* both link local */
         uint8_t candidate_scope = _get_scope(ptr);
         if (candidate_scope == dst_scope) {
             DEBUG("winner for rule 2 (same scope) found\n");
-            winner_set[i] += RULE_2A_PTS;
+            winner_set[i] += RULE_2_PTS;
         }
         else if (candidate_scope < dst_scope) {
             DEBUG("winner for rule 2 (smaller scope) found\n");
-            winner_set[i] += RULE_2B_PTS;
+            winner_set[i] += candidate_scope;
```

Comment:
```suggestion
            winner_set[i] += candidate_scope - dst_scope;
```

Reply:
Doesn't this invert the scoring for scopes. With that the lesser scope gets more precedence.

Reply:
And that is how I understand it should be. I understand `Rule 2` as choose the scope closest to D that is still greater than D.

```
Rule 2: Prefer appropriate scope.
If Scope(SA) < Scope(SB): If Scope(SA) < Scope(D), then prefer SB and
otherwise prefer SA.
```

The above states the case:

- `Scope(SA) < Scope(D) < Scope(SB)` choose SB
- `Scope(D) < Scope(SA) < Scope(SB)` choose SA

Reply:
Yes, but with you suggestion it becames

```
Rule 2: Prefer appropriate scope.
If Scope(SA) > Scope(SB): If Scope(SA) < Scope(D), then prefer SB and
otherwise prefer SA.
```

Reply:
(as opposed to `Scope(SA) < Scope(SB)`)

Reply:
>     * `Scope(SA) < Scope(D) < Scope(SB)` choose SB

Also according to rule 2 "`Scope(SA) < Scope(SB) < Scope(D)` choose SB" btw.

Reply:
`Also according to rule 2 "Scope(SA) < Scope(SB) < Scope(D) choose SB" btw.`

yes you are right, sorry I got confused by reading doc on `dest` and `source` address selection.

But Lesser scope should still be prefered in any other case than `Scope(SA) < Scope(SB) < Scope(D)` right?

Reply:
Okay, let's follow this through. Let's say there are three addresses `SA`, `SB`, and `SC` assigned to the interface and the scopes are as follows:

```
Scope(SA) < Scope(D) < Scope(SB) < Scope(SC)
```

(i know this is not possible with unicast addresses, but let's go through this anyway, to make sure we got this right)

Accoding to rule 2 `SB` should be selected.

If we go through it as proposed by you then for `SA` this line is skipped as `candidate_scope < dst_scope`, for `SB` `winner_set[pos(SB)]` is incremented by `RULE_2_PTS - Scope(SB)`, selecting `SB` for now. For `SC` `winner_set[pos(SC)]` is incremented by `RULE_2_PTS - Scope(SC)` which does not overtake `winner_set[pos(SB)]` as `(RULE_2_PTS - Scope(SB)) > RULE_2_PTS - Scope(SC)`.

I think you are right.

Reply:
@miri64 doesn't the rule mean (easier to map with code hahaha):

```
if (A < B) {
    if( A < D) {
        choose B
    }
}
else {
    chose A
}
```